File: clients/ios/AwareChat-iOS/AwareChat-iOS/Core/Protocol/ProtocolCodec.swift
import Foundation

nonisolated enum ProtocolConstants {
  static let version = 1
}

nonisolated enum ClientEvent: Equatable, Sendable {
  case identify(user: UserDTO)
  case sendMessage(message: MessageDTO)
  case messagePersisted(messageId: UUID)
}

nonisolated extension ClientEvent: Encodable {
  private enum CodingKeys: String, CodingKey {
    case type
    case protocolVersion
    case user
    case message
    case messageId
  }

  func encode(to encoder: any Encoder) throws {
    var container = encoder.container(keyedBy: CodingKeys.self)
    try container.encode(ProtocolConstants.version, forKey: .protocolVersion)
    switch self {
    case .identify(let user):
      try container.encode("identify", forKey: .type)
      try container.encode(user, forKey: .user)
    case .sendMessage(let message):
      try container.encode("send_message", forKey: .type)
      try container.encode(message, forKey: .message)
    case .messagePersisted(let messageId):
      try container.encode("message_persisted", forKey: .type)
      try container.encode(WireValidation.normalized(messageId), forKey: .messageId)
    }
  }
}

nonisolated struct ProtocolErrorEvent: Equatable, Sendable {
  let messageId: String?
  let error: ServerError
}

nonisolated enum ServerEvent: Equatable, Sendable {
  case identityAccepted(user: UserDTO)
  case syncCompleted(pendingCount: Int)
  case messageAccepted(messageId: UUID, serverReceivedAt: Date)
  case incomingMessage(message: MessageDTO)
  case protocolError(ProtocolErrorEvent)
}

nonisolated struct ProtocolCodec: Sendable {
  func encode(_ event: ClientEvent) throws -> Data {
    try JSONEncoder().encode(event)
  }

  func decodeServerEvent(from data: Data) throws -> ServerEvent {
    let decoder = JSONDecoder()
    let header = try decoder.decode(EventHeader.self, from: data)
    guard header.protocolVersion == ProtocolConstants.version else {
      throw WireModelError.invalidProtocolVersion
    }

    switch header.type {
    case "identity_accepted":
      let envelope = try decoder.decode(IdentityAcceptedEnvelope.self, from: data)
      return .identityAccepted(user: envelope.user)
    case "sync_completed":
      let envelope = try decoder.decode(SyncCompletedEnvelope.self, from: data)
      guard envelope.pendingCount >= 0 else { throw WireModelError.invalidPendingCount }
      return .syncCompleted(pendingCount: envelope.pendingCount)
    case "message_accepted":
      let envelope = try decoder.decode(MessageAcceptedEnvelope.self, from: data)
      return .messageAccepted(
        messageId: try WireValidation.uuid(from: envelope.messageId),
        serverReceivedAt: try WireDateCodec.decode(envelope.serverReceivedAt)
      )
    case "incoming_message":
      let envelope = try decoder.decode(IncomingMessageEnvelope.self, from: data)
      guard envelope.message.serverReceivedAt != nil else {
        throw WireModelError.missingServerTimestamp
      }
      return .incomingMessage(message: envelope.message)
    case "protocol_error":
      let envelope = try decoder.decode(ProtocolErrorEnvelope.self, from: data)
      return .protocolError(
        ProtocolErrorEvent(messageId: envelope.messageId, error: envelope.error)
      )
    default:
      throw DecodingError.dataCorrupted(
        .init(codingPath: [], debugDescription: "Unsupported server event type")
      )
    }
  }
}

private nonisolated struct EventHeader: Decodable {
  let type: String
  let protocolVersion: Int
}

private nonisolated struct IdentityAcceptedEnvelope: Decodable {
  let user: UserDTO
}

private nonisolated struct SyncCompletedEnvelope: Decodable {
  let pendingCount: Int
}

private nonisolated struct MessageAcceptedEnvelope: Decodable {
  let messageId: String
  let serverReceivedAt: String
}

private nonisolated struct IncomingMessageEnvelope: Decodable {
  let message: MessageDTO
}

private nonisolated struct ProtocolErrorEnvelope: Decodable {
  let messageId: String?
  let error: ServerError
}

File: clients/android/AwareChat-Android/app/src/main/java/com/example/awarechat_android/core/protocol/ProtocolCodec.kt
package com.example.awarechat_android.core.protocol

import java.time.Instant
import java.util.UUID
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerializationException
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

sealed interface ClientEvent {
    data class Identify(val user: UserDto) : ClientEvent
    data class SendMessage(val message: MessageDto) : ClientEvent
    data class MessagePersisted(val messageId: UUID) : ClientEvent
}

data class ProtocolErrorEvent(
    val messageId: String?,
    val error: ServerErrorDto,
)

sealed interface ServerEvent {
    data class IdentityAccepted(val user: UserDto) : ServerEvent
    data class SyncCompleted(val pendingCount: Int) : ServerEvent
    data class MessageAccepted(val messageId: UUID, val serverReceivedAt: Instant) : ServerEvent
    data class IncomingMessage(val message: MessageDto) : ServerEvent
    data class ProtocolError(val event: ProtocolErrorEvent) : ServerEvent
}

class ProtocolCodec(
    private val json: Json = defaultJson,
) {
    fun encode(event: ClientEvent): String = when (event) {
        is ClientEvent.Identify -> json.encodeToString(IdentifyEnvelope(user = event.user))
        is ClientEvent.SendMessage -> json.encodeToString(SendMessageEnvelope(message = event.message))
        is ClientEvent.MessagePersisted -> json.encodeToString(
            MessagePersistedEnvelope(messageId = event.messageId),
        )
    }

    fun decodeServerEvent(text: String): ServerEvent {
        val header = json.decodeFromString<EventHeader>(text)
        if (header.protocolVersion != PROTOCOL_VERSION) {
            throw WireModelException.InvalidProtocolVersion()
        }

        return when (header.type) {
            "identity_accepted" -> ServerEvent.IdentityAccepted(
                json.decodeFromString<IdentityAcceptedEnvelope>(text).user,
            )
            "sync_completed" -> {
                val count = json.decodeFromString<SyncCompletedEnvelope>(text).pendingCount
                if (count < 0) throw WireModelException.InvalidPendingCount()
                ServerEvent.SyncCompleted(count)
            }
            "message_accepted" -> {
                val envelope = json.decodeFromString<MessageAcceptedEnvelope>(text)
                ServerEvent.MessageAccepted(
                    messageId = WireValidation.uuid(envelope.messageId),
                    serverReceivedAt = WireDateCodec.decode(envelope.serverReceivedAt),
                )
            }
            "incoming_message" -> {
                val message = json.decodeFromString<IncomingMessageEnvelope>(text).message
                if (message.serverReceivedAt == null) {
                    throw WireModelException.MissingServerTimestamp()
                }
                ServerEvent.IncomingMessage(message)
            }
            "protocol_error" -> {
                val envelope = json.decodeFromString<ProtocolErrorEnvelope>(text)
                ServerEvent.ProtocolError(
                    ProtocolErrorEvent(messageId = envelope.messageId, error = envelope.error),
                )
            }
            else -> throw SerializationException("Unsupported server event type")
        }
    }

    companion object {
        val defaultJson: Json = Json {
            ignoreUnknownKeys = true
            encodeDefaults = true
            explicitNulls = true
            isLenient = false
            allowSpecialFloatingPointValues = false
        }
    }
}

@Serializable
private data class EventHeader(val type: String, val protocolVersion: Int)

@Serializable
private data class IdentifyEnvelope(
    val type: String = "identify",
    val protocolVersion: Int = PROTOCOL_VERSION,
    val user: UserDto,
)

@Serializable
private data class SendMessageEnvelope(
    val type: String = "send_message",
    val protocolVersion: Int = PROTOCOL_VERSION,
    val message: MessageDto,
)

@Serializable
private data class MessagePersistedEnvelope(
    val type: String = "message_persisted",
    val protocolVersion: Int = PROTOCOL_VERSION,
    @Serializable(with = UuidSerializer::class)
    val messageId: UUID,
)

@Serializable
private data class IdentityAcceptedEnvelope(val user: UserDto)

@Serializable
private data class SyncCompletedEnvelope(val pendingCount: Int)

@Serializable
private data class MessageAcceptedEnvelope(
    val messageId: String,
    val serverReceivedAt: String,
)

@Serializable
private data class IncomingMessageEnvelope(val message: MessageDto)

@Serializable
private data class ProtocolErrorEnvelope(
    val messageId: String? = null,
    val error: ServerErrorDto,
)
